### PR TITLE
force rollout on input change managed by a different controller

### PR DESCRIPTION
**Pull request by deads2k**
This watches for changes to a configmap and can cause a rollout of the deployment

/assign @mfojtik @sttts

### CODE REVIEW

**Comment by deads2k** [comment]
@enj

**Comment by deads2k** [comment]
@mfojtik now rebased to make it smaller.  does this make sense?  It was @bparees idea to include the hashes in an already tracked object.  I think it works pretty well.

**Comment by deads2k** [comment]
```
   stderr:
    panic: boringcrypto: not in FIPS mode
```

that's fun

/retest

**Comment by deads2k** [comment]
@mfojtik now green.  We need something like this to do cert rotation easily.

File: pkg/operator/detectinputchange/as_configmap.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetConfigMapHash(obj *corev1.ConfigMap) (string, error) {
```
Comment (sttts) [inline pkg/operator/detectinputchange/as_configmap.go]: godocs

File: pkg/operator/detectinputchange/as_configmap.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+func GetSecretHash(obj *corev1.Secret) (string, error) {
+	hasher := fnv.New32()
+	_, err := hasher.Write([]byte(WriteSecretV1OrDie(obj)))
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(hasher.Sum(nil)), nil
+}
+
+func GetStringMapDataHashesForObjects(objs ...runtime.Object) (map[string]string, error) {
```
Comment (sttts) [inline pkg/operator/detectinputchange/as_configmap.go]: MultipleObjectHashStringMap

File: pkg/operator/detectinputchange/as_configmap.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		default:
+			return nil, fmt.Errorf("%v is not handled", objRef.Resource)
+		}
+	}
+
+	return GetStringMapDataHashesForObjects(objs...)
+}
+
+// TODO move to library-go
+func WriteConfigMapV1OrDie(obj *corev1.ConfigMap) string {
```
Comment (sttts) [inline pkg/operator/detectinputchange/as_configmap.go]: write?

File: pkg/operator/detectinputchange/as_configmap.go (hunk 1 above)
Comment (mfojtik) [reply pkg/operator/detectinputchange/as_configmap.go]: also looks like something that should live in library-go

File: pkg/operator/detectinputchange/as_configmap.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetConfigMapHash(obj *corev1.ConfigMap) (string, error) {
+	hasher := fnv.New32()
+	_, err := hasher.Write([]byte(WriteConfigMapV1OrDie(obj)))
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: if _, err := hasher...

File: pkg/operator/detectinputchange/as_configmap.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(hasher.Sum(nil)), nil
+}
+
+func GetSecretHash(obj *corev1.Secret) (string, error) {
+	hasher := fnv.New32()
+	_, err := hasher.Write([]byte(WriteSecretV1OrDie(obj)))
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: nit: if _, err := hasher...

File: pkg/operator/detectinputchange/as_configmap.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+	for _, obj := range objs {
+		switch t := obj.(type) {
+		case *corev1.ConfigMap:
+			hash, err := GetConfigMapHash(t)
+			if err != nil {
+				return nil, err
+			}
+			// this string coercion is lossy, but it should be fairly controlled and must be an allowed name
+			ret[fmt.Sprintf("%s.%s.%s", t.Namespace, t.Name, "configmap")] = hash
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: nit: move `fmt.Sprintf("%s.%s.%s", t.Namespace, t.Name, "configmap")` to function

File: pkg/operator/detectinputchange/as_configmap.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+		case *corev1.Secret:
+			hash, err := GetSecretHash(t)
+			if err != nil {
+				return nil, err
+			}
+			// this string coercion is lossy, but it should be fairly controlled and must be an allowed name
+			ret[fmt.Sprintf("%s.%s.%s", t.Namespace, t.Name, "secret")] = hash
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: nit: replace with function

File: pkg/operator/detectinputchange/as_configmap.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		default:
+			return nil, fmt.Errorf("%T is not handled", t)
+		}
+	}
+
+	return ret, nil
+}
+
+// ObjectReference can be used to reference a particular resource.  Not all group resources are respected by all methods.
+type ObjectReference struct {
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: why not use v1.ObjectReference ?

File: pkg/operator/detectinputchange/as_configmap.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	Resource  schema.GroupResource
+	Namespace string
+	Name      string
+}
+
+func GetStringMapDataHashesForObjectReferences(client kubernetes.Interface, objRefs ...ObjectReference) (map[string]string, error) {
+	objs := []runtime.Object{}
+
+	for _, objRef := range objRefs {
+		switch objRef.Resource {
```
Comment (mfojtik) [inline pkg/operator/detectinputchange/as_configmap.go]: nit: objRef.Resource.Resource ? then you don't need to make the schema

File: pkg/operator/detectinputchange/as_configmap.go (hunk 8 above)
Comment (deads2k) [reply pkg/operator/detectinputchange/as_configmap.go]: > why not use v1.ObjectReference ?

Too many fields that don't make sense.  We've moved away from using it in APIs

**Comment by deads2k** [comment]
updated

**Comment by bparees** [comment]
How does this compare/contrast with resourceapply.SyncConfigMap ?

I'm trying to solve the use case of a configmap in a different namespace (openshift-config) that i need to copy into the api-server namespace on change, and then rollout the apiserver.

I suppose the two pieces can just be combined, ie use the syncconfigmap to copy it, and this new code to trigger the rollout when the local CM gets updated by the sync?

**Comment by deads2k** [comment]
> I suppose the two pieces can just be combined, ie use the syncconfigmap to copy it, and this new code to trigger the rollout when the local CM gets updated by the sync?

Yes, they are separate,but can be combined

**Pull request merged**